README.md
Common TS Readme
================

# What is it 
Common TS is a Pure Typoscript Extension trying to Speed up Your Typo3 Development Dramaticaly
* by giving you a Sane Set of Default Configuration for CONFIG and PAGE object.
* using some automatic Template file Detection to automaticaly include some files like favicon or print.css
* Giving you Copy templates for Commonly Used Typoscript Objects Like Menus.
* all in an easy CSS styleable way by the use classes and ID's, prefixed with CTS- to avoid conflicts
* support for Diffrent Templating Aproaches Classic / Templavoila / MondernBE Grid Layouts
* Search Engine Friendly HTML
* support for XHTML / and HTML5

How To Use 
----------
* Install the extension via Extension manager
* Include the static Extension template directly after css_styled_content (important!)
* check out the Constants Editor for COMMON TS configuration Options

why should the commonTS Static Template be included directly after css_styled_content
for convinicen Common_TS strips the HTML Content Element of the div-wrap around it (Configuration Option). 
therefore it needes to be loaded after css_styled content. also Common TS pedefines a Global config and page object (also configurationable)
but some extension write directly to the "page" object and if the commonTS.page object would be copied after such an event.
the written data would be lost and you get a hard time debugging it. therefore keep it right after css_styled_content and don't worry :)

example Typoscript
------------------

### ul/li Menu 3 Layers Deep

This Example shows you how to get an 3 Level ul/li menu with full classes for every itemstate first and last item.
so the menu is easily stylable with css. also the menu gives you some search engine friendly options

    # page object is definied via commonTS and ready to use
    page.10 =TEMPLATE
    page.10 {
      template = FILE
      tempalte.file = fileadmin/template/template.html
      # Copy the Common TS Template for a Fullfledged menu
      marks.MAINMENU < lib.commonTS.menu.ul
      marks.MAINMENU {
        # Creating menu level 2+3
        2 < .1
        3 < .1
        # give the first layer of the Menu an ID otherwise it would be a plain <ul>|</ul>
        1.wrap = <ul id="main-menu">|</ul>
      }
    }

### Changing an an element-style based on a Page id
    .my-element {background:yellow} 
    #cts-page-16 {background:red}

Configuration Option (Constants Editor)
=======================================
In The constants Editor you find Many Options which help you to Configure Most important options in an Easy way

*** To Be Written ***
CommonTS Structur

all common TS Objects are defined in a namespace consisting of 3 parts
prefix.section.objectdefinition

the prefix is always lib.commonTS 
and is manly there for conflict avoidance

the section indicates a scope of object found in this sup tree.
like menu. contains all menu Templates
there are currently 2 Exceptions to this rule (config and page)

the objectdefinition is the Copy template for the Object.

Automagic FileLoding
--------------------
CommonTS Predifines some files Which will be loaded if present for speedly development:
imagefiles:
images/favicon*
- if found the favicon/shortcuticon for the site is set automaticaly

the following CSS files will be loaded:
  css/reset.css  
  - if you use an CSS reset file

  css/rte.css  
  - ths should be your Rich Text editor Stylesheed so all Classes Defined there are automaticly present in Frontend as welll
  css/layout.css
  - the main Layout of your site
  css/print.css
  - the print layout of your site of course with media="print" set.

Shothandle Constants
--------------------

there are some short handle constants you can use in in your Templates for easy access to most common path
these are configurable via the extension manager

    # Path to the Template Directory
    cts.tpl = fileadmin/template/
    # Path to the Javascript Directory
    cts.js = fileadmin/template/javascript/
    # Path to the CSS Directory
    cts.css = fileadmin/template/css/
    # Path to the Images Directory
    cts.img = fileadmin/template/images/

example usage in your Typoscript Template could be the following

    page.includeCSS.10 = {$cts.css}myCustom.css
    page.includeJS.10 = {$cts.js}extrafunction.js

if you use these Shorthand variables throughout your templates you could easly switch to a new template 
just by changing the constants path. this gets very usfull if your site gets contouinuly developed

Templates Explained
===================

Get a look for your self 
------------------------
i recommend to take a look at the CommonTS code with the Template analyser and Typoscript Explorere
for your better understanding of the workings of this libary

config 
------
*** To Be Written ***

page 
----
*** To Be Written ***

menu.ul 
-------
*** To Be Written ***

menu.ulSimple
-------------
*** To Be Written ***

menu.ulMinimal
-------------- 
*** To Be Written ***




Possibble Extensions
====================
These Options are currently in Disccution for inclusion
  * Print Page Setup
  * Cleaner Setup Structure by moving config and page objects 
  * additonal Bodyclasses (for easy branch Seperation Based on Constants)

Known Problems
==============
currently no Problems are known

ChangeLog

Wed Dec 10 08:49:44 UTC 2014 Christian Wolff <cwolff@aer.de>
RELEASE: Version 1.7.2

Wed Dec 10 09:49:29 2014 +0100 Christian Wolff <cwolff@aer.de> 
bugfix: inreases changed constraints keys dependandcies -> depends

Wed Dec 10 08:23:31 UTC 2014 Christian Wolff <cwolff@aer.de>
RELEASE: Version 1.7.1

13-05-2013 Christian Wolff <chris@connye.com>
added cts-pid-{field:uid} cts-pos-{register:count_MENUOBJ}  to highlight individual positions of menu items

12-08-30 Christian Wolff <chris@connye.com>
  version 1.3.0
	renamed static template file you have to reinclude the static template if you upgrade from a previous version
	Cleaner Body cObject for Better Readability
	Added Automagic Templating Constants 
	enabled css & JS minify+compression+concatination

12-04-19 Christian Wolff <chris@connye.com>
	Version 1.1.0
	addet Option to Setup the Page Object by default via static templates
	and some other options to the Constants
	

12-01-02 Christian Wolff <chris@connye.com>
	Version 1.0.5
	Fixed "modern BE" Backend Layout modus

11-05-10 Christian Wolff <chris@connye.com>
	Version 1.0.3
	Added language class to body tag. cts-lang-<int> 
	now its possible to change background images in CSS
	without switching stylesheets.
	
10-03-08 Christian Wolff  <chris@connye.com>

	* Initial release


res/demopage/js/jquery-htmlBeautify.js
// taken from http://www.davidpirek.com/blog/html-beautifier-jquery-plugin
(function($){

    var level=0,
     lSize=100,
        finishTabifier = function(code, targetElm) {
	        code=code.replace(/\n\s*\n/g, '\n');  //blank lines
	        code=code.replace(/^[\s\n]*/, ''); //leading space
	        code=code.replace(/[\s\n]*$/, ''); //trailing space

	        targetElm.val(code);

	        level=0;

        },
        cleanHTML = function(code, targetElm) {
        
	        var i=0;
	        function cleanAsync() {
		        var iStart=i;
		        for (; i'!=code.charAt(point)) point++;
			        tag=code.substr(start, point-start);
			        i=point;

			        //if this is a special tag, deal with it!
			        if ('!--'==tag.substr(1,3)) {
				        if (!tag.match(/--$/)) {
					        while ('-->'!=code.substr(point, 3)) point++;
					        point+=2;
					        tag=code.substr(start, point-start);
					        i=point;
				        }
				        if ('\n'!=out.charAt(out.length-1)) out+='\n';
				        out+=tabs();
				        out+=tag+'>\n';
			        } else if ('!'==tag[1]) {
				        out=placeTag(tag+'>', out);
			        } else if ('?'==tag[1]) {
				        out+=tag+'>\n';
			        } else if (t=tag.match(/^<(script|style)/i)) {
				        t[1]=t[1].toLowerCase();
				        tag=cleanTag(tag);
				        out=placeTag(tag, out);
				        end=String(code.substr(i+1)).toLowerCase().indexOf('');
	        return tag;
        },
        ownLine=['area', 'body', 'head', 'hr', 'i?frame', 'link', 'meta', 'noscript', 'style', 'table', 'tbody', 'thead', 'tfoot'],
        contOwnLine=['li', 'dt', 'dt', 'h[1-6]', 'option', 'script'],
        lineBefore = new RegExp(
	        '^<(/?'+ownLine.join('|/?')+'|'+contOwnLine.join('|')+')[ >]'
        ),
        lineAfter=new RegExp('^<(br|/?'+ownLine.join('|/?')+'|/'+contOwnLine.join('|/')+')[ >]'),
        newLevel=['blockquote', 'div', 'dl', 'fieldset', 'form', 'frameset','map', 'ol', 'p', 'pre', 'select', 'td', 'th', 'tr', 'ul'],
        newLevel=new RegExp('^]'),
        placeTag = function(tag, out) {
	        var nl=tag.match(newLevel);
	        if (tag.match(lineBefore) || nl) {
		        out=out.replace(/\s*$/, '');
		        out+="\n";
	        }

	        if (nl && '/'==tag.charAt(1)) level--;
	        if ('\n'==out.charAt(out.length-1)) out+=tabs();
	        if (nl && '/'!=tag.charAt(1)) level++;

	        out+=tag;
	        if (tag.match(lineAfter) || tag.match(newLevel)) {
		        out=out.replace(/ *$/, '');
		        out+="\n";
	        }
	        return out;
        };


    // plugin definition
    $.fn.htmlBeautifier = function(code, settings) {

        // iterate and reformat each matched element
        return this.each(function() {
        
            cleanHTML(code, $(this));
            
        });
    };

})(jQuery);

ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "commonts".
 *
 * Auto generated 17-05-2013 12:57
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Common TS',
	'description' => 'Library of Typoscript object to get your Site Development a BOOST',
	'category' => 'fe',
	'shy' => 0,
	'version' => '1.7.2',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'beta',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Christian Wolff',
	'author_email' => 'chris@connye.com',
	'author_company' => '',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.5.0-6.2.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:20:{s:9:"ChangeLog";s:4:"3b17";s:12:"ext_icon.gif";s:4:"b303";s:14:"ext_tables.php";s:4:"6f6e";s:9:"README.md";s:4:"775e";s:10:"todo.tasks";s:4:"7694";s:14:"doc/manual.sxw";s:4:"0fa0";s:19:"doc/wizard_form.dat";s:4:"6ebc";s:20:"doc/wizard_form.html";s:4:"5db9";s:21:"res/demopage/demo1.ts";s:4:"1944";s:21:"res/demopage/demo2.tx";s:4:"1749";s:21:"res/demopage/demo3.ts";s:4:"a7bb";s:21:"res/demopage/demo4.ts";s:4:"ebb5";s:23:"res/demopage/layout.css";s:4:"d41d";s:27:"res/demopage/css/layout.css";s:4:"a7a4";s:23:"res/demopage/js/init.js";s:4:"5561";s:38:"res/demopage/js/jquery-htmlBeautify.js";s:4:"a33a";s:25:"static/base/constants.txt";s:4:"dbe3";s:21:"static/base/setup.txt";s:4:"47c7";s:25:"static/demo/constants.txt";s:4:"eeb0";s:21:"static/demo/setup.txt";s:4:"abc8";}',
	'suggests' => array(
	),
);

?>